packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m

```diff
@@ -232,52 +258,106 @@ @interface FLTMarkersController ()
-                            registrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+- (instancetype)initWithClusterManagersController:
+                    (nullable FLTClusterManagersController *)clusterManagers
+                                          channel:(FlutterMethodChannel *)channel
+                                          mapView:(GMSMapView *)mapView
+                                        registrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+  _clusterManagersController = clusterManagers;
   self = [super init];
   if (self) {
-    _methodChannel = methodChannel;
+    _methodChannel = channel;
     _mapView = mapView;
     _markerIdentifierToController = [[NSMutableDictionary alloc] init];
     _registrar = registrar;
   }
   return self;
 }
 
+- (instancetype)initWithMethodChannel:(FlutterMethodChannel *)methodChannel
+                              mapView:(GMSMapView *)mapView
+                            registrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+  return [self initWithClusterManagersController:nil
+                                         channel:methodChannel
+                                         mapView:mapView
+                                       registrar:registrar];
+}
+
 - (void)addMarkers:(NSArray *)markersToAdd {
   for (NSDictionary *marker in markersToAdd) {
-    CLLocationCoordinate2D position = [FLTMarkersController getPosition:marker];
-    NSString *identifier = marker[@"markerId"];
-    FLTGoogleMapMarkerController *controller =
-        [[FLTGoogleMapMarkerController alloc] initMarkerWithPosition:position
-                                                          identifier:identifier
-                                                             mapView:self.mapView];
-    [controller interpretMarkerOptions:marker registrar:self.registrar];
-    self.markerIdentifierToController[identifier] = controller;
+    [self addMarker:marker];
   }
 }
 
+- (void)addMarker:(NSDictionary *)markerToAdd {
+  id markerIdentifier = markerToAdd[@"markerId"];
+  id clusterManagerIdentifier = markerToAdd[@"clusterManagerId"];
+  CLLocationCoordinate2D position = [FLTMarkersController getPosition:markerToAdd];
+  GMSMarker *marker = [GMSMarker markerWithPosition:position];
+  FLTGoogleMapMarkerController *controller =
+      [[FLTGoogleMapMarkerController alloc] initWithMarker:marker
+                                          markerIdentifier:markerIdentifier
+                                  clusterManagerIdentifier:clusterManagerIdentifier
+                                                   mapView:self.mapView];
+  [controller interpretMarkerOptions:markerToAdd registrar:self.registrar];
+  if (clusterManagerIdentifier && [clusterManagerIdentifier isKindOfClass:[NSString class]]) {
+    id clusterManager =
+        [_clusterManagersController clusterManagerWithIdentifier:clusterManagerIdentifier];
+    if (marker && [clusterManager isKindOfClass:[GMUClusterManager class]]) {
```

Comment:
do we need type check here? can the `_clusterManagersController` here directly return `GMUClusterManager` type?

Reply:
changed to null check